connmgr/connmanager.go

```diff
@@ -255,39 +276,75 @@ out:
 			case handleDisconnected:
 				connReq, ok := conns[msg.id]
 				if !ok {
-					connReq, ok = pendingConns[msg.id]
-					if ok && !msg.retry {
-						connReq.updateState(ConnFailed)
-
-						log.Debugf("Cancelling: %v", connReq)
-						delete(pendingConns, msg.id)
-						return
+					connReq, ok = pending[msg.id]
+					if !ok {
+						log.Errorf("Unknown connid=%d",
+							msg.id)
+						continue
 					}
-				}
 
-				if connReq != nil {
-					connReq.updateState(ConnDisconnected)
+					// Pending connection was found, remove
+					// it from pending map if we should
+					// ignore a later, successful
+					// connection.
+					connReq.updateState(ConnCanceled)
+					log.Debugf("Canceling: %v", connReq)
+					delete(pending, msg.id)
+					continue
+
+				} else {
+					// An existing connection was located,
+					// mark as disconnected and execute
+					// disconnection callback.
+					log.Debugf("Disconnected from %v",
+						connReq)
+					delete(conns, msg.id)
+
 					if connReq.conn != nil {
 						connReq.conn.Close()
 					}
-					log.Debugf("Disconnected from %v", connReq)
-					delete(conns, msg.id)
 
 					if cm.cfg.OnDisconnection != nil {
 						go cm.cfg.OnDisconnection(connReq)
 					}
+				}
 
-					if uint32(len(conns)) < cm.cfg.TargetOutbound && msg.retry {
-						cm.handleFailedConn(connReq)
-					}
-				} else {
-					log.Errorf("Unknown connection: %d", msg.id)
+				// All internal state has been cleaned up, if
+				// this connection is being removed, we will
+				// make no further attempts with this request.
+				if !msg.retry {
+					connReq.updateState(ConnDisconnected)
+					continue
+				}
+
+				// Otherwise, we will attempt a reconnection if
+				// we do not have enough peers, or if this is a
+				// persistent peer. The connection request is
+				// readded to the pending map, so that
+				// subsequent processing of connections and
+				// failures do not ignore the request.
+				if uint32(len(conns)) < cm.cfg.TargetOutbound ||
+					connReq.Permanent {
+
+					connReq.updateState(ConnPending)
+					log.Debugf("Reconnecting to %v",
+						connReq)
+					pending[msg.id] = connReq
+					cm.handleFailedConn(connReq)
 				}
 
 			case handleFailed:
 				connReq := msg.c
-				connReq.updateState(ConnFailed)
-				log.Debugf("Failed to connect to %v: %v", connReq, msg.err)
+
+				if _, ok := pending[connReq.id]; !ok {
+					log.Debugf("Ignoring connection for "+
+						"canceled conn req: %v", connReq)
```

Comment:
Shouldn't we ensure that it isn't still in the regular connection map?

Reply:
I think this would only be possible if a user calls `Connect(cr)` on a connection request that has already connected, but then the `Dial` to the same address fails. In all other cases, the connection is moved from `conns -> pending` before executing `handleFailedConn`. In terms of side-effects, this should be fine, since the request will be ignored and we're already connected. That being said, we could query the connection request for its state before attempting the `Dial` and ignore as appropriate?